File: 442. Find All Duplicates in an Array/findDuplicates.py
from typing import List

class Solution:
  def findDuplicates(self, nums: List[int]) -> List[int]:
    res = []
    
    for num in nums:
      n = abs(num)
      if nums[n - 1] < 0:
        res.append(n)
        
      nums[n - 1] = -nums[n - 1]
      
    return res


nums = [4, 3, 2, 7, 8, 2, 3, 1]
solution = Solution()
print('findDuplicates: ', solution.findDuplicates(nums))

File: 442. Find All Duplicates in an Array/findDuplicates.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const findDuplicates = (nums) => {
  const res = [];

  for (let i = 0; i < nums.length; i++) {
    let num = Math.abs(nums[i]);
    if (nums[num - 1] < 0) {
      res.push(num);
    }

    nums[num - 1] = -nums[num - 1]
  }

  return res;
};

const nums = [4, 3, 2, 7, 8, 2, 3, 1];
console.log('findDuplicates: ', findDuplicates(nums));
